File: web/src/services/passwordValidationService.ts
export interface PasswordValidationRule {
	id: string;
	test: (password: string) => boolean;
	messageKey: string;
	isValid?: boolean;
}
export interface PasswordValidationResult {
	isValid: boolean;
	rules: PasswordValidationRule[];
	score: number;
}

const passwordValidationRules: PasswordValidationRule[] = [
	{
		id: "minLength",
		test: (password: string) => password.length >= 8,
		messageKey: "auth.password.validation.minLength",
	},
	{
		id: "hasUppercase",
		test: (password: string) => /[A-Z]/.test(password),
		messageKey: "auth.password.validation.hasUppercase",
	},
	{
		id: "hasLowercase",
		test: (password: string) => /[a-z]/.test(password),
		messageKey: "auth.password.validation.hasLowercase",
	},
	{
		id: "hasNumber",
		test: (password: string) => /\d/.test(password),
		messageKey: "auth.password.validation.hasNumber",
	},
	{
		id: "hasSymbol",
		test: (password: string) =>
			/[!@#$%^&*()_+\-=[\]{};':"\\|,.<>/?]/.test(password),
		messageKey: "auth.password.validation.hasSymbol",
	},
];

function getPasswordStrength(
	score: number,
): "weak" | "fair" | "good" | "strong" {
	if (score < 40) return "weak";
	if (score < 60) return "fair";
	if (score < 80) return "good";
	return "strong";
}

export const PasswordValidationService = {
	validatePassword(password: string): PasswordValidationResult {
		const validatedRules = passwordValidationRules.map((rule) => ({
			...rule,
			isValid: rule.test(password),
		}));
		const validRulesCount = validatedRules.filter(
			(rule) => rule.isValid,
		).length;
		const isValid = validRulesCount === passwordValidationRules.length;
		const score = Math.round(
			(validRulesCount / passwordValidationRules.length) * 100,
		);
		return {
			isValid,
			rules: validatedRules,
			score,
		};
	},
	isPasswordValid(password: string): boolean {
		return PasswordValidationService.validatePassword(password).isValid;
	},
	getPasswordStrength,
	getPasswordStrengthColor(score: number): string {
		const strength = getPasswordStrength(score);
		switch (strength) {
			case "weak":
				return "text-red-600";
			case "fair":
				return "text-orange-500";
			case "good":
				return "text-yellow-500";
			case "strong":
				return "text-green-600";
			default:
				return "text-gray-500";
		}
	},
	getPasswordStrengthBgColor(score: number): string {
		const strength = getPasswordStrength(score);
		switch (strength) {
			case "weak":
				return "bg-red-500";
			case "fair":
				return "bg-orange-500";
			case "good":
				return "bg-yellow-500";
			case "strong":
				return "bg-green-500";
			default:
				return "bg-gray-300";
		}
	},
};

File: backend/api/src/services/password_validation_service.py
"""
Service de validation de mot de passe pour le backend.
Implémente les mêmes règles que le frontend pour assurer la cohérence.
"""
import re


class PasswordValidationRule:
    """Règle de validation de mot de passe."""
    def __init__(self, rule_id: str, test_func, message: str):
        self.id = rule_id
        self.test = test_func
        self.message = message
class PasswordValidationResult:
    """Résultat de la validation d'un mot de passe."""
    def __init__(self, is_valid: bool, failed_rules: list[str]):
        self.is_valid = is_valid
        self.failed_rules = failed_rules
class PasswordValidationService:
    """Service de validation de mot de passe côté backend."""
    RULES = [
        PasswordValidationRule(
            "min_length",
            lambda password: len(password) >= 8,
            "Le mot de passe doit contenir au moins 8 caractères"
        ),
        PasswordValidationRule(
            "has_uppercase",
            lambda password: re.search(r'[A-Z]', password) is not None,
            "Le mot de passe doit contenir au moins une lettre majuscule"
        ),
        PasswordValidationRule(
            "has_lowercase",
            lambda password: re.search(r'[a-z]', password) is not None,
            "Le mot de passe doit contenir au moins une lettre minuscule"
        ),
        PasswordValidationRule(
            "has_number",
            lambda password: re.search(r'\d', password) is not None,
            "Le mot de passe doit contenir au moins un chiffre"
        ),
        PasswordValidationRule(
            "has_symbol",
            lambda password: re.search(r'[!@#$%^&*()_+\-=\[\]{};\':"\\|,.<>\/?]', password) is not None,
            "Le mot de passe doit contenir au moins un symbole"
        )
    ]
    @classmethod
    def validate_password(cls, password: str) -> PasswordValidationResult:
        """
        Valide un mot de passe selon toutes les règles définies.
        Args:
            password: Le mot de passe à valider
        Returns:
            PasswordValidationResult: Résultat de la validation
        """
        failed_rules = []
        for rule in cls.RULES:
            if not rule.test(password):
                failed_rules.append(rule.message)
        is_valid = len(failed_rules) == 0
        return PasswordValidationResult(is_valid, failed_rules)
    @classmethod
    def is_password_valid(cls, password: str) -> bool:
        """
        Vérifie si le mot de passe respecte toutes les règles (validation simple).
        Args:
            password: Le mot de passe à valider
        Returns:
            bool: True si le mot de passe est valide, False sinon
        """
        return cls.validate_password(password).is_valid
    @classmethod
    def get_validation_errors(cls, password: str) -> list[str]:
        """
        Retourne la liste des erreurs de validation pour un mot de passe.
        Args:
            password: Le mot de passe à valider
        Returns:
            List[str]: Liste des messages d'erreur
        """
        return cls.validate_password(password).failed_rules